In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/arrests/Arrests.csv


# **Unsupervised Learning**

**Custering**

> Técnica de mineração de dados para agrupar dados não identificados por similaridade ou diferença

**Association**

> Usa regras diferentes para encontrar relação entre variáveis em qualquer dataset fornecido
>
> Frequentemente usado em análise de compra e mecanismos de recomendação, por isso também é chamado de sistema de recomendação
> > "Clientes que compraram isso geralmente compram esse..."

**Dimensionality Reduction**

> Técnica usada quando o número de funcionalidade ou dimensões do dataset é muito alto
> > Reduz o número de entradas para o tamanho gerenciável mantendo a integridade dos dados


**Supervised Learning (Aprendizado Supervisionado) é ideal para:**
* Detecção de spam
* Análise de sentimento
* Previsão do tempo
* Previsão de preços



**Unsupervised Learning (Aprendizado Não Supervisionado) é ideal para:**
* Mecanismo de recomendação
* Personalidade de cliente
* Imagens médica

# **K-Means Clustering**

* Determina o número de grupos e usa esse valor como parametro
 * Os valores de um grupo devem ser o mais similares
 * Os valores não devem ser totalmente iguais
 

**Passos para entendimento**

1. Determinar o centro do agrupamento
2. Classificar amostras fora do centro de acordo com a distância
3. Determinar o novo centro (mudar do centro antigo para o novo) de acordo com a classificação
4. Repete os passos 2 e 3 até ser estável

* Esse método determina o valor de K
 * A soma do quadrado da distancia dos pontos até o centro do grupo é calculada de acordo com cada valor de K


**Muito usado em...**

* Classificação de documentos
* Localização de crimes
* Análise de desempenho de jogadores
* Segmentação de clientes
* Cálculo de custo médio de clientes de seguro
* Pesquisa de terremotos
* Segmentação de venda (mais usado)

# **Importing the Libraries**

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [3]:
df = pd.read_csv("/kaggle/input/arrests/Arrests.csv")
df.head()

Unnamed: 0  Murder  Assault  UrbanPop  Rape
0     Alabama    13.2      236        58  21.2
1      Alaska    10.0      263        48  44.5
2     Arizona     8.1      294        80  31.0
3    Arkansas     8.8      190        50  19.5
4  California     9.0      276        91  40.6

In [4]:
df = pd.read_csv("/kaggle/input/arrests/Arrests.csv", index_col = 0)
df.head()

Murder  Assault  UrbanPop  Rape
Alabama       13.2      236        58  21.2
Alaska        10.0      263        48  44.5
Arizona        8.1      294        80  31.0
Arkansas       8.8      190        50  19.5
California     9.0      276        91  40.6

Murder - % de homicídios por 100.000 pessoas (13,2 = 13.200 de 100.000 casos são homicídios)

Assault = Assaltos ocorridos a cada 100.000 ocorrências ()

UrbanPop = % da população urbana

Rape = Estupros cometidos a cada 100.000 casos

# **Getting to know the dataset**

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50 entries, Alabama to Wyoming
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Murder    50 non-null     float64
 1   Assault   50 non-null     int64  
 2   UrbanPop  50 non-null     int64  
 3   Rape      50 non-null     float64
dtypes: float64(2), int64(2)
memory usage: 2.0+ KB


In [6]:
df.describe()

Murder     Assault   UrbanPop       Rape
count  50.00000   50.000000  50.000000  50.000000
mean    7.78800  170.760000  65.540000  21.232000
std     4.35551   83.337661  14.474763   9.366385
min     0.80000   45.000000  32.000000   7.300000
25%     4.07500  109.000000  54.500000  15.075000
50%     7.25000  159.000000  66.000000  20.100000
75%    11.25000  249.000000  77.750000  26.175000
max    17.40000  337.000000  91.000000  46.000000

# **Hopkins Test**

In [7]:
#!pip install pyclustertend

In [8]:
#from pyclustering import hopkins
from sklearn.preprocessing import scale

In [9]:
#hopkins(scale(df), df.shape[0])